README.markdown

```diff
@@ -392,6 +392,34 @@ You can find real world examples of `Doctest` being used below:
   * [base Data/Maybe.hs](https://github.com/ghc/ghc/blob/669cbef03c220de43b0f88f2b2238bf3c02ed64c/libraries/base/Data/Maybe.hs#L36-L79)
   * [base Data/Functor.hs](https://github.com/ghc/ghc/blob/669cbef03c220de43b0f88f2b2238bf3c02ed64c/libraries/base/Data/Functor.hs#L34-L64)
 
+## Test dependencies
+By default `doctest` runs all test in the same GHCi session. This means some
+operations, e.g. `:set -XTypeApplications`, carry over to the new module creating
+a potentially unexpected dependency between tests. In order to isolate tests,
+you can use `--isolate-modules`.
+
+Note that by default each module (and its dependencies) will be recompiled for
+each GHCi sessions. If you're using cabal-doctest consider using
+`--use-package-db` to load functions from a precompiled package.
+
+## Parallel tests
+Run tests in parallel with `-jN`. Either `-j` or `-j0` sets the number of threads
+to `GHC.Conc.numCapabilities`. The default is `-j1`. Any other `N` implies
+`N` threads and `--isolate-modules`. Make sure to compile your doctest binary
+with `-threaded -with-rtsopts=-N`. For example:
+
+```
+test-suite doctests
+  type:             exitcode-stdio-1.0
+  default-language: Haskell2010
+  main-is:          doctests.hs
+  ghc-options:      -Wall -Wcompat -threaded -with-rtsopts=-N
+```
+
+
+## Use precompiled code to run tests
+When using `cabal-doctest`, `doctest` has access to precompiled function
+definitions. To use them, pass in `--use-package-db`.
```

Comment:
Is it ever useful to *not* have `--use-package-db`? If it isn't, let's avoid adding an option and just have doctest be smarter.

Reply:
It depends on how a project is setup. It's possible to just run the `doctest` binary on a file like:

```
doctest Foo.hs
```

In that case, the definitions in `Foo.hs` aren't in any package db. This changes when you use something like `cabal-doctest` which first compiles the library and then passes in a `-package my-package` flag to doctest. It also passes in `-isrc` (if the tests are located in `src/`) to allow `doctest` to discover them. If `ghci` is started with the same flags, it'll prefer to recompile modules from `src/` over looking for them in its package database. There's no way to tell `ghci` to prefer precompiled definitions AFAIK.

In any case, `--use-package-db` would probably break all test suites _not_ setup using `cabal-doctest`, so I'd be hesitant to _not_ make it an option.

Reply:
Scrap that, it should be as simple as:

```
import Module
```

if it fails, use

```
:load Module
```

instead.

Reply:
I'm a bit confused. [The GHCi manual says this](https://downloads.haskell.org/ghc/latest/docs/html/users_guide/ghci.html#ghci-compiled):

> When loading up source modules with `:load`, GHCi normally looks for any corresponding compiled object files, and will use one in preference to interpreting the source if possible.

That sounds like, if doctest used `:load`, we'd be getting the desired behaviour with no further changes.

What am I missing here? Is it something to do with modules within the current package vs modules from dependencies, or something to do with doctest being dumb about package databases managed by cabal-install/Stack and so GHC not knowing where to find them, or something else again?

Reply:
@quasicomputational There's basically two ways I could do this:

1. Try and detect whether to load from package db or local files at the very start, then apply the detected setting for all tests. (We could add a flag `--use-package-db={always,never,auto}`.)
2. For every module try `import Module`. If that fails use `:load Module`.

I'm thinking (1) is the best approach. Do you have preferences?

Reply:
> What am I missing here? Is it something to do with modules within the current package vs modules from dependencies, or something to do with doctest being dumb about package databases managed by cabal-install/Stack and so GHC not knowing where to find them, scoping of non-exported top-level definitions, or something else again?

I believe what it's trying to say is that it will try to load from a local object file (`*.o`) if present. So if you do `:load Test` it'll look for `Test.o` first. It doesn't try to load from a package though:

```
Prelude> :l Data.List

<no location info>: error:
    module ‘Data.List’ cannot be found locally
Failed, no modules loaded.
Prelude> import Data.List
```

Reply:
> or something to do with doctest being dumb about package databases managed by cabal-install/Stack and so GHC not knowing where to find them

There's not much `doctest` can do to be "smart" about package databases. With the `v2-*` cabal commands there's no way to find out where package databases are located. `cabal-doctest` works around this by doing CPP/Cabal magic to provide the necessary flags to `doctest` during compilation. For example, a typical set of arguments may look like (copying from my dayjob's project):

```
-i/home/martijn/code/clash-compiler/clash-prelude/src
[..]
-XTypeOperators
-XViewPatterns
-XNoStarIsType
[..]
-package-db=/home/martijn/.cabal/store/ghc-8.10.3/package.db
-package-db=/home/martijn/code/clash-compiler/dist-newstyle/packagedb/ghc-8.10.3
-package-db=/home/martijn/code/clash-compiler/dist-newstyle/build/x86_64-linux/ghc-8.10.3/clash-prelude-1.3.0/package.conf.inplace
[..]
-package-id=array-0.5.4.0
[..]
-package=clash-prelude-1.3.0
[..]
Clash.Annotations.TopEntity
Clash.Annotations.Primitive
[..]
```

These flags get passed to GHCi, which can use `-i/home/martijn/code/clash-compiler/clash-prelude/src` to try and locally compile modules. But it could also use the various `-package-db` and `-package=clash-prelude-1.3.0` to find precompiled definitions for the functions under test.